rts/lra/lra-proxy/api/src/main/java/io/narayana/lra/client/internal/proxy/nonjaxrs/LRAParticipant.java

```diff
@@ -0,0 +1,249 @@
+            case 0:
+                result = invokeMethod(type, method, participant);
+                break;
+            case 1:
+                result = invokeMethod(type, method, participant, lraId);
+                break;
+            case 2:
+                result = invokeMethod(type, method, participant, lraId, parentId);
+                break;
+            default:
+                throw new InvalidLRAParticipantDefinitionException(
+                    method.toGenericString() + ": invalid number of arguments: " + method.getParameterCount());
+        }
+
+        return processResult(result, lraId, method, type);
+    }
+
+    private Object invokeMethod(String type, Method method, Object o, Object... args) {
+        try {
+            return method.invoke(o, args);
+        } catch (InvocationTargetException e) {
+            if (e.getTargetException() instanceof WebApplicationException) {
+                return ((WebApplicationException) e.getTargetException()).getResponse();
+            }
+
+            // other exceptions differ based on the participant method type
+            if (type.equals(COMPENSATE)) {
+                return ParticipantStatus.FailedToCompensate.name();
+            } else if (type.equals(COMPLETE)) {
+                return ParticipantStatus.FailedToComplete.name();
+            } else {
+                // @Status and @Forget should return HTTP 500
+                return Response.status(Response.Status.INTERNAL_SERVER_ERROR).build();
+            }
+        } catch (IllegalAccessException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    private Response processResult(Object result, URI lraId, Method method, String type) {
+        Response.ResponseBuilder builder = Response.status(Response.Status.OK);
+
+        if (result instanceof CompletionStage) {
+            // store the CompletionStage result and respond compensating / completing
+            participantStatusMap.put(lraId, Optional.empty());
+            ((CompletionStage<?>) result)
+                .thenAccept(res -> participantStatusMap.replace(lraId, Optional.ofNullable(res)))
+                .exceptionally(throwable -> {
+                    participantStatusMap.replace(lraId, Optional.ofNullable(throwable));
+                    return null;
+                });
+            return builder.status(Response.Status.ACCEPTED).build();
+        }
+
+        if (method.getReturnType().equals(Void.TYPE)) {
+            // void return type and no exception was thrown
+            builder.entity(type.equals(COMPLETE) ? ParticipantStatus.Completed.name() : ParticipantStatus.Compensated.name());
+        } else if (result == null) {
+            builder.status(Response.Status.NOT_FOUND);
+        } else if (result instanceof ParticipantStatus) {
+            builder.entity(((ParticipantStatus) result).name());
+        } else if (result instanceof Response) {
+            return (Response) result;
+        } else {
```

Comment:
This is where you would end up if the participant business logic throws an exception.